tensorflow/python/ops/special_math_ops.py

```diff
@@ -324,13 +324,43 @@ def sort_key(input_index, a):
     num_summed_elements = _total_size(t0_shape[-len(axes_to_sum):])
     new_shape = t0_shape[:len(preserved_axes)] + (num_broadcast_elements_t0,
                                                   num_summed_elements)
+    # Make new_shape valid, if invalid to be used in reshape
+    none_dims = [i for i, d in enumerate(new_shape)
+                 if d is None or d == -1]
+    if len(none_dims) > 1:
+      new_shape = list(new_shape)
+      for i in none_dims[:-1]:
+        if i < len(preserved_axes):
+          new_shape[i] = array_ops.shape(t0)[i]
+        elif i == len(new_shape) - 2:
+          new_shape[i] = math_ops.reduce_prod(
+            array_ops.shape(t0)[len(preserved_axes):-len(axes_to_sum)])
+        else:
+          new_shape[i] = math_ops.reduce_prod(
+            array_ops.shape(t0)[-len(axes_to_sum):])
+      new_shape = tuple(new_shape)
     t0 = _reshape_if_necessary(t0, new_shape)
 
     t1_shape = tuple(x.value for x in t1.get_shape())
     num_broadcast_elements_t1 = _total_size(
         t1_shape[len(preserved_axes)+len(axes_to_sum):])
     new_shape = t1_shape[:len(preserved_axes)] + (num_summed_elements,
                                                   num_broadcast_elements_t1)
+    # Make new_shape valid, if invalid to be used in reshape
+    none_dims = [i for i, d in enumerate(new_shape)
+                 if d is None or d == -1]
+    if len(none_dims) > 1:
```

Comment:
it would be nice to have more comments about what it means to 'make new_shape valid' in order to guide the code reader about what is being done here.